### PR TITLE
[Bug] replace GamePlayKit with aStar for path finding

### LINKED ISSUE #16: pathfinding error occasionally
# Bug description

Base on the DemoApp, I extended the graph like this

**Graph declaration** https://github.com/ampm/MappaMundi/blob/ampm/reproduce_finpath_bug/UITests/DemoMappaMundi.swift

image

The shortest way from screen `itemDetail` to `foo_5` should be:

`itemDetail`->`item_action_1`->`foo_5`.

But,  it doesn't work all the time, sometimes, it found out a longer path.

I added some code to the `MMNavigator.swift` for debugging and invoked the navigation testing for 10 times.
```
    override func invokeTest() {
        for time in 0...10 {
            print("this test is invoking: \(time) times")
            super.invokeTest()
        }
    }
    func testReproduceFindpathError() {
        navigator.performAction(Actions.addItem)
        navigator.goto(Screens.itemDetail)
        navigator.goto(Screens.foo5)
    }
```

The log shows:
image

# Troubleshooting

At first, I doubted if it's a bug of Apple's GameplayKit ?  I created a Playground to verify it.

In that Playground, I copied the graph in UITests and run it for 100 times.

image

But, the results can't prove my guess, it always returns the correct path.

# How to reproduce this bug?
pls feel free to check out this repo 
https://github.com/ampm/MappaMundi
**branch name**: reproduce_finpath_bug
then you can run the case  `func testReproduceFindpathError()` in `DemoUITests.swift`.

@jhugman  Can you give me any clue then I'll try to fix it?

Thanks a lot.

### PR DESCRIPTION
**Pull request by ampm**
Hi @garvankeeley @jhugman ,  can you help to review this PR and to see if it is possible to merge it back into master branch?
# Checkpoints
- [x] have tested it and now the pathfinding works stable.
- [x] Test with FireFox-iOS:  SmokeTest works as well as with `GamePlayKit`(except of `testAccountManagementPage`, this case failed because of the code itself).

# Test Result
image

# Description 
This PR has resolved #16 

-  **Remove GamePlayKit**
-   **Replace it with aStar** 
The implementation of aStar algorithm is borrowed from @Dev1an https://github.com/Dev1an/A-Star (MIT license) 

# Rewind

One more thing I want to share with you is, in this branch(for reproducing the bug, if you remember)

https://github.com/ampm/MappaMundi/blob/ampm/reproduce_finpath_bug/

We repeatedly run `testReproduceFindpathError` ten times with the following code, and our app will relaunch before each invokes, we can reproduce the error easily, the right or wrong result comes alternatively. 

 ```
 override func invokeTest() {
        for time in 0...10 {
            print("this test is invoking: \(time) times")
            super.invokeTest()
        }
    }
    func testReproduceFindpathError() {
        navigator.nowAt(Screens.itemDetail)
        navigator.goto(Screens.foo5)
    }
```

But if we repeat it in this way(we run the test manually),  
- if the path is incorrect at the first loop, it will show us with 10 mistakes.
- in a similar way,  it will show us with 10 right results consistently if the path is right from begging.

 ```
func testReproduceFindpathError() {
        for _ in 0...10 {
            navigator.nowAt(Screens.itemDetail)
            navigator.goto(Screens.foo5)
        }
    }
```

### CODE REVIEW

**Comment by Dev1an** [comment]
Hello @ampm
Nice to hear that you are using my A* implementation. If you are interested, I made an SPM package that you can now easily depend upon using Xcode 11.

I also updated the code to use the new hashing function.
SPM dependency

**Comment by garvankeeley** [comment]
Hopefully we can upgrade to Xcode 11 in the coming months, but we want this project to be compatible with the build macs for Firefox iOS which is currently limited to Xcode 10.

**Comment by garvankeeley** [comment]
Can you change the inclusion of A* to use:
Cartfile: 
```
github "Dev1an/A-Star" "master"
```
Assuming 'master' is the working version here.
and add the Cartfile.resolved

Then include Carthage/Checkouts/A-Star/[the needed files] swift files in the mappamundi project as they are currently.

File: MappaMundi/aStar/MMNode.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+//
+//  MMNode.swift
+//  MappaMundi
+//
+//  Created by Yang, Roy on 2019/6/7.
+//  Copyright © 2019 Mozilla Corporation. All rights reserved.
```
Comment (garvankeeley) [inline MappaMundi/aStar/MMNode.swift]: Please use the mozilla licence header, thanks

**Comment by Dev1an** [comment]
I added a version tag so you can safely use all versions compatible with version 1.0.0
Cartfile:
```
github "Dev1an/A-Star" ~> 2.0.0
```

File: MappaMundi/aStar/MMNode.swift (hunk 1 above)
Comment (ampm) [reply MappaMundi/aStar/MMNode.swift]: sure, updated.

**Comment by ampm** [comment]
> Can you change the inclusion of A* to use:
> Cartfile:
> 
> ```
> github "Dev1an/A-Star" "master"
> ```
> 
> Assuming 'master' is the working version here.
> and add the Cartfile.resolved
> 
> Then include Carthage/Checkouts/A-Star/[the needed files] swift files in the mappamundi project as they are currently.

Hi @garvankeeley , have integrated with Carthage.

**Comment by ampm** [comment]
Hi @garvankeeley, have updated code, pls help to review it again, thanks.
Really appreciate @Dev1an's help.

**Comment by ampm** [comment]
hi @garvankeeley, pls on hold, I need to do more test to make sure it works well with FireFox-iOS.

**Comment by ampm** [comment]
FYI, @garvankeeley, these two cases `testConditionalEdges()`  and `testNavigatorActions`

failed at `XCTAssertFalse(navigator.can(goto: Screens.itemDetail))`.

This is the graph, we indeed can go from `ItemList` to `ItemDetail`,  I don't know why we write that assertion.  
image

In my opinion, the test design is wrong, they should fail. 

How about let's remove these two `XCTAssertFalse(navigator.can(goto: Screens.itemDetail))`?

**Comment by garvankeeley** [comment]
I'll run all the Firefox iOS tests with this today and see how it goes.

**Comment by ampm** [comment]
> I'll run all the Firefox iOS tests with this today and see how it goes.

sure, but pls let me know before merging, because of the Astar source code in MappamMundi is a little different with the version specified in Carhtage/Carthage.resolve files. (I copied them instead of using the checked out source at this moment.)

I have created a PR for it and waiting for @Dev1an to review it. https://github.com/Dev1an/A-Star/pull/7 . once it is merged, I can update the Carthage file in this PR and link checked files into this project.

**Comment by garvankeeley** [comment]
I'll wait until your PR is merged. For now, the test results are looking good on Firefox iOS.

**Comment by garvankeeley** [comment]
Mind if I ping for a status update on this PR?

**Comment by ampm** [comment]
Hi @garvankeeley, sorry for the late response.

I have integrated it with Carthage,  but I'm stuck with the nested framework in the output. Can you help to figure it out?

image

**Comment by ampm** [comment]
hi @garvankeeley , it's ready to merge now.

To apply this update in FireFox-iOS, I can create another PR with following action~~s~~ for it.

~~1. add the dependency of~~
~~`github "Dev1an/A-Star"    ~> 3.0.0-beta-1`~~
2. add the build output `AStar.framework` into the link binaries build phase of 
- `L10nSnapshotTests`
- `XCUITests`


Thanks a lot.

**Comment by garvankeeley** [comment]
I'll land this and then update Firefox iOS.

**Pull request merged**